# Pandas
* Which is a build in Library and using Pandas heavily for data manipulation, visualisation, building machine learning models, etc.
* There are two main data structures in Pandas - Series and Dataframes

### The pandas Series (similar to a numpy array)

* A dataframe is simply a table where each column is a pandas series. 
    * Creating a pandas series
    * Indexing series
    
### Dataframes

* Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having       an index and columns having meaningful names.

    * Creating dataframes from dictionaries
    * Importing CSV data files as pandas dataframes
    * Reading and summarising dataframes
    * Sorting dataframes
    
[Birds_dataset](https://drive.google.com/file/d/1A5Z-uNxmKLljstrX8j-1j2ydThYbe6JU/view?usp=sharing)

### Creating Pandas Series
* A dataframe is simply a table where each column is a pandas series.
* 1D-Array

In [1]:
import pandas as pd

In [4]:
# converting list into series
s1 = pd.Series([11,22,33,44,55])
print(type(s1))
s1

<class 'pandas.core.series.Series'>


0    11
1    22
2    33
3    44
4    55
dtype: int64

In [5]:
# converting numpy array into series
import numpy as np
n = np.array([123,345,456,567])
s2 = pd.Series(n)
s2

0    123
1    345
2    456
3    567
dtype: int32

In [6]:
s3 = pd.Series(['sai','hema','prabhakar'])
s3

0          sai
1         hema
2    prabhakar
dtype: object

In [9]:
# Date series
s4 = pd.date_range(start = '2020-06-22',end = '2020-07-11')
len(s4)
s4

DatetimeIndex(['2020-06-22', '2020-06-23', '2020-06-24', '2020-06-25',
               '2020-06-26', '2020-06-27', '2020-06-28', '2020-06-29',
               '2020-06-30', '2020-07-01', '2020-07-02', '2020-07-03',
               '2020-07-04', '2020-07-05', '2020-07-06', '2020-07-07',
               '2020-07-08', '2020-07-09', '2020-07-10', '2020-07-11'],
              dtype='datetime64[ns]', freq='D')

In [10]:
# converting Dict to Series
d = {'a': 10 ,'t' : 100, 'u' : 1000}
s5 = pd.Series(d)
s5

a      10
t     100
u    1000
dtype: int64

In [17]:
# index starts from 1 to 10 and data in series are power of index value
s6 = pd.Series(np.arange(1,11)**2,index = range(1,11))
s6

1       1
2       4
3       9
4      16
5      25
6      36
7      49
8      64
9      81
10    100
dtype: int32

```NOTE``` : **Make sure number of elements in the index is always equal to number of elements specified in the series**

In [18]:
s7 = pd.Series(['2334','abc',1343],index = ['a',13,24.56])
s7

a        2334
13        abc
24.56    1343
dtype: object

In [19]:
s7.index

Index(['a', 13, 24.56], dtype='object')

In [20]:
# indexing and slicing
s7['a']

'2334'

In [22]:
s7[-1:]

24.56    1343
dtype: object

In [23]:
s7[1:2]

13    abc
dtype: object

In [26]:
s7[0:3:2]

a        2334
24.56    1343
dtype: object

## Pandas Dataframe
* Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

* Creating dataframes from dictionaries

* There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV     files, etc.

-- | col1 | col2 | col3
-- | ---|---|---
r1 | 1 | 2 | 3
r2 | 4 | 5 | 6
r3 | 7 | 8 | 9



*  Reading and Summarising Dataframes
*  Cleaning DataFrame
*  Sorting 
*  Indexing and Selecting  Data
*  Concat Data
*  Export Data 

In [28]:
# create dataframe using dict
d = {
    "Std_name" : ["Chaitanya",'Ameer','sai','geetha'],
    "Python"   : [89,67,98,59]    
}
df = pd.DataFrame(d)
df

Std_name  Python
0  Chaitanya      89
1      Ameer      67
2        sai      98
3     geetha      59

In [29]:
# Create df using List
li  = [['eng',100],['Maths',90],['physics',67],['social',80]]
df1 = pd.DataFrame(li)
df1

0    1
0      eng  100
1    Maths   90
2  physics   67
3   social   80

In [30]:
df1.columns = ['Subject','Marks']
df1

Subject  Marks
0      eng    100
1    Maths     90
2  physics     67
3   social     80

In [34]:
# Explicitly specified Index
df1.index = ['a','b','c','d']
df1

Subject  Marks
a      eng    100
b    Maths     90
c  physics     67
d   social     80

In [53]:
d = {
    "Std_name" : pd.Series(["Chaitanya",'Ameer','sai','geetha'],index = ["a",'b','c','d']),
    "Python_Marks"   : pd.Series([89,67,98],index = ['a','b','c'] ) 
}
df2 = pd.DataFrame(d)
df2
# NaN -- Not a Number

Std_name  Python_Marks
a  Chaitanya          89.0
b      Ameer          67.0
c        sai          98.0
d     geetha           NaN

In [66]:
d = {'std_name' : pd.Series(['chaitanaya','ameer','sai','geetha'],index = ['a','b','c','d']), 
     'python_Marks' : pd.Series([89,67,98,59],index = ['a','b','c','d'])
    }
df2 = pd.DataFrame(d)
df2


std_name  python_Marks
a  chaitanaya            89
b       ameer            67
c         sai            98
d      geetha            59

In [56]:
d={
    
    "state":pd.Series(["ap","tm","kerala"],index=[11,22,33]),
    "cases":pd.Series([100,200],index=[11,22])

}

s=pd.DataFrame(d)
s


state  cases
11      ap  100.0
22      tm  200.0
33  kerala    NaN

In [57]:
df2.isnull().sum()

Std_name        0
Python_Marks    1
dtype: int64

In [58]:
df2.index

Index(['a', 'b', 'c', 'd'], dtype='object')

In [59]:
df2.columns

Index(['Std_name', 'Python_Marks'], dtype='object')

In [60]:
df2.shape  # (rows,columns)

(4, 2)

In [61]:
df2.values

array([['Chaitanya', 89.0],
       ['Ameer', 67.0],
       ['sai', 98.0],
       ['geetha', nan]], dtype=object)

In [63]:
##explicity  specified index
df1.index = range(1,11)
df1

ValueError: Length mismatch: Expected axis has 4 elements, new values have 10 elements

In [75]:
# If u want to set index as one column 
#df2 = df2.set_index('std_name')
df2.set_index('std_name',inplace = True)


KeyError: 'std_name'

In [76]:
df2

python_Marks
std_name                
chaitanaya            89
ameer                 67
sai                   98
geetha                59

In [127]:
# Reading data from CSV file

data = pd.read_csv('birds.csv')
data

id    huml   humw   ulnal  ulnaw   feml  femw    tibl  tibw   tarl  \
0      0   80.78   6.68   72.01   4.88  41.81  3.70    5.50  4.03  38.70   
1      1   88.91   6.63   80.53   5.59  47.04  4.30   80.22  4.51  41.50   
2      2   79.97   6.37   69.26   5.28  43.07  3.90   75.35  4.04  38.31   
3      3   77.65   5.70   65.76   4.77  40.04  3.52   69.17  3.40  35.78   
4      4   62.80   4.84   52.09   3.73  33.95  2.72   56.27  2.96  31.88   
5      5   61.92   4.78   50.46   3.47  49.52  4.41   56.95  2.73  29.07   
6      6   79.73   5.94   67.39   4.50  42.07  3.41   71.26  3.56  37.22   
7      7   86.98   5.68   74.52   4.55  44.46  3.78   76.02  3.81  37.94   
8      8  118.20   7.82  116.64   6.13  59.33  5.45  110.00  5.58  61.62   
9      9  145.00  10.42  144.00   7.05  70.96  7.44  120.00  7.31  78.67   
10    10  165.00  11.45  156.00   8.68  80.25  7.85  143.00  8.25  86.61   
11    11  186.00   9.83  152.00   8.76  56.02  7.02  185.00  8.07  90.80   
12    12  172.00   8.44  142.00   8.43  52.18  6.68  180.00  9.62  82.58   
13    13  148.91   6.78  121.35   6.50  44.29  6.33  155.00  6.68  73.71   
14    14  149.19   6.98  121.48   6.21  43.33  5.39  155.00  6.61  75.65   
15    15  140.59   6.59  115.97   5.81  39.75  5.61  166.00  7.83  74.86   
16    16  135.23   6.22  108.78   5.83  37.99  5.13  157.00  7.18  72.44   
17    17  124.53   6.01  100.37   5.82  34.54  5.43  145.00  7.12  65.92   
18    18  127.03   6.18  102.53   5.64  36.52  4.89  150.00  6.78  67.27   
19    19  106.02   4.47   95.79   4.33  41.96  4.80  129.59  6.94  70.29   
20    20  113.84   4.80  100.71   4.39  43.95  5.08  133.27  7.41  70.91   
21    21  108.62   5.04   97.70   4.66  45.08  5.78  128.00  7.14  65.48   
22    22  110.33   5.86   97.10   4.60  43.82  5.73  126.00  7.86  64.17   
23    23   63.62   3.42   59.00   2.75  30.16  2.92   74.66  4.38  39.26   
24    24   70.73   3.69   65.19   2.93  32.52  3.54   80.42  4.36  43.70   
25    25   71.65   3.37   66.11   2.77  36.98  3.06   77.11  3.94  38.04   
26    26   47.51   7.23   32.49   8.26  50.29  5.00   75.20  4.95  22.51   
27    27   46.20   7.06   32.73   8.58  52.15  5.11   75.99  4.77  23.24   
28    28   41.13   6.71   31.00   7.73  49.51  4.69   71.94  4.54  22.46   
29    29   72.50  14.78   45.83  10.33  77.02  7.76  128.75  6.49  32.48   
..   ...     ...    ...     ...    ...    ...   ...     ...   ...    ...   
390  390   20.43   1.59   22.68   1.39  20.00  1.54   35.63  1.33  25.57   
391  391   22.19   1.86   24.93   1.52  24.14  1.78   40.17  1.73  29.61   
392  392   23.27   2.00   24.80   1.83  26.04  1.79   41.91  1.67  29.76   
393  393   18.97   1.58   21.26   1.40  21.50  1.55   36.52  1.52  25.04   
394  394   15.99   1.35   18.32   1.13  15.84  1.19   30.05  1.23  22.94   
395  395   15.62   1.28   18.52   1.06  15.75  1.17   28.63  1.03  21.39   
396  396   16.51   1.47   20.56   1.43  15.88  1.27     NaN  1.19  17.63   
397  397   16.80   1.58   19.77   1.39  16.86  1.29   29.75  1.18  21.77   
398  398   15.55   1.48   18.93   1.30  17.58  1.23   31.65  1.14  24.12   
399  399   23.32   2.06   27.90   1.84  22.49  1.66   34.77  1.42  23.05   
400  400   20.41   1.82   22.66   1.71  18.14  1.52   29.30  1.21  20.29   
401  401   21.53   1.91   25.30   1.67  20.80  1.65   34.08  1.33  23.86   
402  402   22.84   2.21   27.70   2.06  22.13  1.72   34.60  1.69  25.08   
403  403   20.95   1.89   24.12   1.66  21.24  1.84   36.71  1.53  25.99   
404  404   20.36   1.87   22.19   1.60    NaN  1.77   37.47  1.64  25.54   
405  405   17.30   1.68   20.43   1.33  15.51  1.15   25.96  0.88  17.41   
406  406   16.86   1.58   20.72   1.36  14.92  1.13   24.56  0.98  16.76   
407  407   15.70   1.42   18.58   1.20  15.22  1.08   25.39  0.92  16.88   
408  408   19.54   1.86   23.35   1.47  18.34  1.46   29.05  1.35  19.68   
409  409   23.27   2.15   27.62   2.16  21.86  1.64   35.55  1.51  23.23   
410  410   23.04   2.15   28.29   1.94  21.46  1.67   

In [82]:
data.shape

(420, 12)

In [83]:
data.columns

Index(['id', 'huml', 'humw', 'ulnal', 'ulnaw', 'feml', 'femw', 'tibl', 'tibw',
       'tarl', 'tarw', 'type'],
      dtype='object')

In [84]:
data.values

array([[0, 80.78, 6.68, ..., 38.7, 3.84, 'SW'],
       [1, 88.91, 6.63, ..., 41.5, 4.01, 'SW'],
       [2, 79.97, 6.37, ..., 38.31, 3.34, 'SW'],
       ...,
       [417, 18.79, 1.63, ..., 22.86, 1.21, 'SO'],
       [418, 20.38, 1.78, ..., 25.98, 1.24, 'SO'],
       [419, 17.89, 1.44, ..., 21.69, 1.05, 'SO']], dtype=object)

In [87]:
type(data['type'])

pandas.core.series.Series

In [89]:
type(data[['type']])

pandas.core.frame.DataFrame

In [92]:
len(data['type'].value_counts())

6

In [93]:
data['type'].value_counts()

SO    128
SW    116
W      65
R      50
P      38
T      23
Name: type, dtype: int64

In [94]:
data

id    huml   humw   ulnal  ulnaw   feml  femw    tibl  tibw   tarl  \
0      0   80.78   6.68   72.01   4.88  41.81  3.70    5.50  4.03  38.70   
1      1   88.91   6.63   80.53   5.59  47.04  4.30   80.22  4.51  41.50   
2      2   79.97   6.37   69.26   5.28  43.07  3.90   75.35  4.04  38.31   
3      3   77.65   5.70   65.76   4.77  40.04  3.52   69.17  3.40  35.78   
4      4   62.80   4.84   52.09   3.73  33.95  2.72   56.27  2.96  31.88   
5      5   61.92   4.78   50.46   3.47  49.52  4.41   56.95  2.73  29.07   
6      6   79.73   5.94   67.39   4.50  42.07  3.41   71.26  3.56  37.22   
7      7   86.98   5.68   74.52   4.55  44.46  3.78   76.02  3.81  37.94   
8      8  118.20   7.82  116.64   6.13  59.33  5.45  110.00  5.58  61.62   
9      9  145.00  10.42  144.00   7.05  70.96  7.44  120.00  7.31  78.67   
10    10  165.00  11.45  156.00   8.68  80.25  7.85  143.00  8.25  86.61   
11    11  186.00   9.83  152.00   8.76  56.02  7.02  185.00  8.07  90.80   
12    12  172.00   8.44  142.00   8.43  52.18  6.68  180.00  9.62  82.58   
13    13  148.91   6.78  121.35   6.50  44.29  6.33  155.00  6.68  73.71   
14    14  149.19   6.98  121.48   6.21  43.33  5.39  155.00  6.61  75.65   
15    15  140.59   6.59  115.97   5.81  39.75  5.61  166.00  7.83  74.86   
16    16  135.23   6.22  108.78   5.83  37.99  5.13  157.00  7.18  72.44   
17    17  124.53   6.01  100.37   5.82  34.54  5.43  145.00  7.12  65.92   
18    18  127.03   6.18  102.53   5.64  36.52  4.89  150.00  6.78  67.27   
19    19  106.02   4.47   95.79   4.33  41.96  4.80  129.59  6.94  70.29   
20    20  113.84   4.80  100.71   4.39  43.95  5.08  133.27  7.41  70.91   
21    21  108.62   5.04   97.70   4.66  45.08  5.78  128.00  7.14  65.48   
22    22  110.33   5.86   97.10   4.60  43.82  5.73  126.00  7.86  64.17   
23    23   63.62   3.42   59.00   2.75  30.16  2.92   74.66  4.38  39.26   
24    24   70.73   3.69   65.19   2.93  32.52  3.54   80.42  4.36  43.70   
25    25   71.65   3.37   66.11   2.77  36.98  3.06   77.11  3.94  38.04   
26    26   47.51   7.23   32.49   8.26  50.29  5.00   75.20  4.95  22.51   
27    27   46.20   7.06   32.73   8.58  52.15  5.11   75.99  4.77  23.24   
28    28   41.13   6.71   31.00   7.73  49.51  4.69   71.94  4.54  22.46   
29    29   72.50  14.78   45.83  10.33  77.02  7.76  128.75  6.49  32.48   
..   ...     ...    ...     ...    ...    ...   ...     ...   ...    ...   
390  390   20.43   1.59   22.68   1.39  20.00  1.54   35.63  1.33  25.57   
391  391   22.19   1.86   24.93   1.52  24.14  1.78   40.17  1.73  29.61   
392  392   23.27   2.00   24.80   1.83  26.04  1.79   41.91  1.67  29.76   
393  393   18.97   1.58   21.26   1.40  21.50  1.55   36.52  1.52  25.04   
394  394   15.99   1.35   18.32   1.13  15.84  1.19   30.05  1.23  22.94   
395  395   15.62   1.28   18.52   1.06  15.75  1.17   28.63  1.03  21.39   
396  396   16.51   1.47   20.56   1.43  15.88  1.27     NaN  1.19  17.63   
397  397   16.80   1.58   19.77   1.39  16.86  1.29   29.75  1.18  21.77   
398  398   15.55   1.48   18.93   1.30  17.58  1.23   31.65  1.14  24.12   
399  399   23.32   2.06   27.90   1.84  22.49  1.66   34.77  1.42  23.05   
400  400   20.41   1.82   22.66   1.71  18.14  1.52   29.30  1.21  20.29   
401  401   21.53   1.91   25.30   1.67  20.80  1.65   34.08  1.33  23.86   
402  402   22.84   2.21   27.70   2.06  22.13  1.72   34.60  1.69  25.08   
403  403   20.95   1.89   24.12   1.66  21.24  1.84   36.71  1.53  25.99   
404  404   20.36   1.87   22.19   1.60    NaN  1.77   37.47  1.64  25.54   
405  405   17.30   1.68   20.43   1.33  15.51  1.15   25.96  0.88  17.41   
406  406   16.86   1.58   20.72   1.36  14.92  1.13   24.56  0.98  16.76   
407  407   15.70   1.42   18.58   1.20  15.22  1.08   25.39  0.92  16.88   
408  408   19.54   1.86   23.35   1.47  18.34  1.46   29.05  1.35  19.68   
409  409   23.27   2.15   27.62   2.16  21.86  1.64   35.55  1.51  23.23   
410  410   23.04   2.15   28.29   1.94  21.46  1.67   

In [95]:
data.head(5)  # returns top 5 Rows 

id   huml  humw  ulnal  ulnaw   feml  femw   tibl  tibw   tarl  tarw type
0   0  80.78  6.68  72.01   4.88  41.81  3.70   5.50  4.03  38.70  3.84   SW
1   1  88.91  6.63  80.53   5.59  47.04  4.30  80.22  4.51  41.50  4.01   SW
2   2  79.97  6.37  69.26   5.28  43.07  3.90  75.35  4.04  38.31  3.34   SW
3   3  77.65  5.70  65.76   4.77  40.04  3.52  69.17  3.40  35.78  3.41   SW
4   4  62.80  4.84  52.09   3.73  33.95  2.72  56.27  2.96  31.88  3.13   SW

In [96]:
data.tail(5) # returns last 5 rows

id   huml  humw  ulnal  ulnaw   feml  femw   tibl  tibw   tarl  tarw  \
415  415  17.96  1.63  19.25   1.33  18.36  1.54  31.25  1.33  21.99  1.15   
416  416  19.21  1.64  20.76   1.49  19.24  1.45  33.21  1.28  23.60  1.15   
417  417  18.79  1.63  19.83   1.53  20.96  1.43  34.45  1.41  22.86  1.21   
418  418  20.38  1.78  22.53   1.50  21.35  1.48  36.09  1.53  25.98  1.24   
419  419  17.89  1.44  19.26   1.10  17.62  1.34  29.81  1.24  21.69  1.05   

    type  
415   SO  
416   SO  
417   SO  
418   SO  
419   SO

In [101]:
data.sample(5)  # returns random row

id    huml  humw  ulnal  ulnaw   feml  femw    tibl  tibw   tarl  tarw  \
329  329   19.00  1.68  24.00   1.45  14.86  1.23   27.19  1.01  20.05  0.86   
68    68   38.34  2.37  36.96   2.01  15.12  1.30   35.46  1.27  22.96  1.20   
343  343   30.10  2.49  34.69   2.26  28.74  2.08   49.08  2.10  36.09  1.88   
268  268   27.57  2.67  31.60   2.12  18.92  1.52   30.42  1.58  12.46  1.46   
19    19  106.02  4.47  95.79   4.33  41.96  4.80  129.59  6.94  70.29  3.35   

    type  
329   SO  
68    SW  
343   SO  
268    P  
19    SW

In [103]:
data[['humw','ulnal']].head()

humw  ulnal
0  6.68  72.01
1  6.63  80.53
2  6.37  69.26
3  5.70  65.76
4  4.84  52.09

In [106]:
data.shape

(420, 12)

```NOTE``` : **AXis =1 represents columns and Axis = 0 represents rows**

In [109]:
data.head(3)

id   huml  humw  ulnal  ulnaw   feml  femw   tibl  tibw   tarl  tarw type
0   0  80.78  6.68  72.01   4.88  41.81   3.7   5.50  4.03  38.70  3.84   SW
1   1  88.91  6.63  80.53   5.59  47.04   4.3  80.22  4.51  41.50  4.01   SW
2   2  79.97  6.37  69.26   5.28  43.07   3.9  75.35  4.04  38.31  3.34   SW

In [113]:
data.sort_index(axis = 0).head()

id   huml  humw  ulnal  ulnaw   feml  femw   tibl  tibw   tarl  tarw type
0   0  80.78  6.68  72.01   4.88  41.81  3.70   5.50  4.03  38.70  3.84   SW
1   1  88.91  6.63  80.53   5.59  47.04  4.30  80.22  4.51  41.50  4.01   SW
2   2  79.97  6.37  69.26   5.28  43.07  3.90  75.35  4.04  38.31  3.34   SW
3   3  77.65  5.70  65.76   4.77  40.04  3.52  69.17  3.40  35.78  3.41   SW
4   4  62.80  4.84  52.09   3.73  33.95  2.72  56.27  2.96  31.88  3.13   SW

In [124]:
data.sort_values(by='type',inplace= True,ascending=False)
data['type'].head()
#data['type'].value_counts()

153    W
130    W
173    W
167    W
139    W
Name: type, dtype: object

In [122]:
data['type'].value_counts()

SO    128
SW    116
W      65
R      50
P      38
T      23
Name: type, dtype: int64

In [130]:
# Indexing and Selecting Data
data[100:150]

id    huml   humw   ulnal  ulnaw    feml  femw    tibl  tibw    tarl  \
100  100   67.06   3.54   67.98   3.24   24.82  1.77   51.88  1.86   30.11   
101  101   67.59   3.73   68.55   3.19   24.46  1.89   50.06  2.36   27.21   
102  102  100.27   5.75  107.95   4.86   35.50  3.03   69.54  3.21   39.83   
103  103   68.50   3.77   71.15   3.34   24.86  1.95   51.44  2.10   30.13   
104  104   68.15   3.84   72.31   3.42   24.23  1.90   50.26  2.13   29.16   
105  105  107.76   4.71  110.21   4.12   36.15  3.11   76.18  3.77   49.61   
106  106  109.70   4.65  109.87   4.04   35.16  3.25   76.80  3.56   47.27   
107  107  100.69   4.80  101.28   4.13   36.36  3.36   85.62  4.20   52.54   
108  108  100.08   5.07  100.81   4.04   37.15  3.77   87.55  3.96   52.91   
109  109   70.04   4.74   61.26   3.97   28.12  2.98   71.11  3.44   40.23   
110  110   75.95   5.42   66.16   4.16   30.71  3.10   78.94  3.56   44.82   
111  111  106.23   6.82  104.02   5.07   40.06  4.96  105.34  4.99   57.27   
112  112  105.19   7.10   96.85   5.43   37.61  4.43  102.04  4.90   54.78   
113  113   76.49   5.36   69.06   4.02   30.07  3.03   72.41  3.71   43.11   
114  114   65.66   3.79   61.06   3.35   24.95  2.44   63.76  2.72   40.19   
115  115   98.73   6.48   91.69   5.31   35.34  4.02   92.36  4.46   50.53   
116  116  157.00   9.00  192.00   7.02   85.60  6.44  227.00  6.84  175.00   
117  117  190.00  11.92  225.00   8.55  101.75  7.75  240.00  7.71  175.00   
118  118   70.01   3.71   76.90   2.88   49.42  3.08   79.56  3.35   52.74   
119  119   92.93   5.28  110.94   3.87   55.11  4.16  137.00  4.29   96.37   
120  120  118.52   6.87  134.91   5.06   78.20  5.24  132.00  5.38   85.28   
121  121  120.26   7.01  132.42   5.40   67.00  5.35  130.10  5.22   87.53   
122  122   67.63   4.37   60.17   3.73   53.03  4.08   92.18  4.31   54.38   
123  123   57.08   4.28   46.72   3.63   53.64  4.11   89.93  4.23   58.14   
124  124   48.19   3.44   43.71   3.01   50.46  3.60   94.86  3.45   55.00   
125  125   30.99   2.09   26.14   1.82   32.83  2.09   50.67  1.98   30.80   
126  126   34.17   2.21   27.80   1.97   35.82  2.32   51.77  2.06   33.13   
127  127   37.64   2.20   31.58   1.87   37.69  2.37   57.06  2.19   37.19   
128  128   35.11   2.51   29.18   2.54   23.20  1.81   38.09  1.84   18.33   
129  129   33.37   2.69   30.06   2.51   22.86  1.84   38.30  1.87   19.13   
130  130   34.80   2.44   29.12   2.38   23.16  1.60   35.50  1.51   19.07   
131  131   74.89   5.63   60.54   4.67   45.38  3.30   72.16  3.53   31.94   
132  132   76.71   5.82   61.54   5.20   47.87  3.53   75.24  4.21   35.11   
133  133   45.25   3.50   38.72   3.33   28.60  2.03   46.36  2.39   24.03   
134  134   46.40   3.39   38.60   3.32   29.00  2.36   49.08  2.49   26.55   
135  135   45.68   3.95   38.98   3.51   29.15  2.44   45.23  2.52   24.64   
136  136   45.18   3.44   38.53   3.15   28.31  2.21   46.93  2.36   25.86   
137  137   50.22   4.31   36.33   3.79   25.70  2.24   46.42  2.10   25.87   
138  138   44.71   3.93   32.87   2.97   20.87  1.76   40.79  1.80   23.42   
139  139   84.92   7.90   63.16   5.78   46.77  3.84   82.87  4.36   37.17   
140  140   88.16   8.15   66.99   6.25   46.55  3.86   83.02  3.99   36.71   
141  141   82.39   4.80   92.50   3.64   47.86  4.05  140.00  4.26   95.39   
142  142   83.48   5.06   94.46   3.57   49.26  4.04  110.00  3.80   82.00   
143  143   27.94   1.93   30.25   1.66   19.55  1.52   35.21  1.38   24.47   
144  144   29.40   2.05   32.56   1.90   20.92  1.46   38.96  1.45   23.98   
145  145   68.92   4.48   75.44   3.97   38.49  3.02   89.16  3.53   67.75   
146  146   67.37   4.34   78.50   4.05   27.53  2.53   45.73  2.53   25.57   
147  147   38.79   2.61   46.60   2.64   18.64  1.54   32.09  1.37   15.80   
148  148  116.72   6.08  133.82   6.32   44.55  3.60   78.04  3.99   48.78   
149  149  138.48   7.54  157.00   6.72   62.20  5.07  116.09  5.83   72.49   

     tarw type  
100  1.63 

### iloc  - for accessing rows using integers
### loc -  for accessing rows other than integers indicies

In [131]:
data.iloc[100:120]

id    huml   humw   ulnal  ulnaw    feml  femw    tibl  tibw    tarl  \
100  100   67.06   3.54   67.98   3.24   24.82  1.77   51.88  1.86   30.11   
101  101   67.59   3.73   68.55   3.19   24.46  1.89   50.06  2.36   27.21   
102  102  100.27   5.75  107.95   4.86   35.50  3.03   69.54  3.21   39.83   
103  103   68.50   3.77   71.15   3.34   24.86  1.95   51.44  2.10   30.13   
104  104   68.15   3.84   72.31   3.42   24.23  1.90   50.26  2.13   29.16   
105  105  107.76   4.71  110.21   4.12   36.15  3.11   76.18  3.77   49.61   
106  106  109.70   4.65  109.87   4.04   35.16  3.25   76.80  3.56   47.27   
107  107  100.69   4.80  101.28   4.13   36.36  3.36   85.62  4.20   52.54   
108  108  100.08   5.07  100.81   4.04   37.15  3.77   87.55  3.96   52.91   
109  109   70.04   4.74   61.26   3.97   28.12  2.98   71.11  3.44   40.23   
110  110   75.95   5.42   66.16   4.16   30.71  3.10   78.94  3.56   44.82   
111  111  106.23   6.82  104.02   5.07   40.06  4.96  105.34  4.99   57.27   
112  112  105.19   7.10   96.85   5.43   37.61  4.43  102.04  4.90   54.78   
113  113   76.49   5.36   69.06   4.02   30.07  3.03   72.41  3.71   43.11   
114  114   65.66   3.79   61.06   3.35   24.95  2.44   63.76  2.72   40.19   
115  115   98.73   6.48   91.69   5.31   35.34  4.02   92.36  4.46   50.53   
116  116  157.00   9.00  192.00   7.02   85.60  6.44  227.00  6.84  175.00   
117  117  190.00  11.92  225.00   8.55  101.75  7.75  240.00  7.71  175.00   
118  118   70.01   3.71   76.90   2.88   49.42  3.08   79.56  3.35   52.74   
119  119   92.93   5.28  110.94   3.87   55.11  4.16  137.00  4.29   96.37   

     tarw type  
100  1.63   SW  
101  1.77   SW  
102  3.28   SW  
103  2.37   SW  
104  2.05   SW  
105  3.35   SW  
106  2.98   SW  
107  3.03   SW  
108  2.82   SW  
109  1.91   SW  
110  2.22   SW  
111  3.20   SW  
112  2.83   SW  
113  2.08   SW  
114  1.89   SW  
115  2.69   SW  
116  6.77    W  
117  7.00    W  
118  3.02    W  
119  3.79    W

In [132]:
data.head(5)

id   huml  humw  ulnal  ulnaw   feml  femw   tibl  tibw   tarl  tarw type
0   0  80.78  6.68  72.01   4.88  41.81  3.70   5.50  4.03  38.70  3.84   SW
1   1  88.91  6.63  80.53   5.59  47.04  4.30  80.22  4.51  41.50  4.01   SW
2   2  79.97  6.37  69.26   5.28  43.07  3.90  75.35  4.04  38.31  3.34   SW
3   3  77.65  5.70  65.76   4.77  40.04  3.52  69.17  3.40  35.78  3.41   SW
4   4  62.80  4.84  52.09   3.73  33.95  2.72  56.27  2.96  31.88  3.13   SW

In [134]:
data['type'][100:110]

100    SW
101    SW
102    SW
103    SW
104    SW
105    SW
106    SW
107    SW
108    SW
109    SW
Name: type, dtype: object

In [140]:
data.loc[100:200,['type','huml']]

type    huml
100   SW   67.06
101   SW   67.59
102   SW  100.27
103   SW   68.50
104   SW   68.15
105   SW  107.76
106   SW  109.70
107   SW  100.69
108   SW  100.08
109   SW   70.04
110   SW   75.95
111   SW  106.23
112   SW  105.19
113   SW   76.49
114   SW   65.66
115   SW   98.73
116    W  157.00
117    W  190.00
118    W   70.01
119    W   92.93
120    W  118.52
121    W  120.26
122    W   67.63
123    W   57.08
124    W   48.19
125    W   30.99
126    W   34.17
127    W   37.64
128    W   35.11
129    W   33.37
..   ...     ...
171    W   22.89
172    W   64.75
173    W   35.97
174    W   71.35
175    W   73.44
176    W   33.20
177    W   32.31
178    W   28.51
179    W   34.12
180    W   87.01
181    T   43.97
182    T   44.58
183    T   20.25
184    T   48.70
185    T   35.74
186    T   37.69
187    T   31.26
188    T   30.97
189    T   30.31
190    T   32.02
191    T   32.34
192    T   30.88
193    T   32.11
194    T   32.61
195    T   34.63
196    T   32.28
197    T   92.82
198    T   55.62
199    T   32.68
200    T   34.24

[101 rows x 2 columns]

# Concat

In [145]:
d1 ={
    
    "state":pd.Series(["ap","tm","kerala"],index=[11,22,33]),
    "cases":pd.Series([100,200,300],index=[11,22,33])

}
d2 ={
    
    "state":pd.Series(["a","t","k"],index=[11,22,33]),
    "cases":pd.Series([10,20,30],index=[11,22,33])

}
df3 = pd.DataFrame(d1)
df4 = pd.DataFrame(d2)

In [146]:
pd.concat([df3,df4],axis=1)  # column concat

state  cases state  cases
11      ap    100     a     10
22      tm    200     t     20
33  kerala    300     k     30

In [151]:
df3.append(df4)

state  cases
11      ap    100
22      tm    200
33  kerala    300
11       a     10
22       t     20
33       k     30

In [148]:
d_concat = pd.concat([df3,df4],axis=0)  # row concat
d_concat

state  cases
11      ap    100
22      tm    200
33  kerala    300
11       a     10
22       t     20
33       k     30

### To Export data

In [150]:
d_concat.to_csv('data_Concat.csv')

In [156]:
n1 = np.array([1,2,3,4,5])
n2 = np.array([1,2,3,4,5])
n1-n2
len(n1)
n1[1]

2

In [158]:
for i in range(len(n1)):
    print(n1[i]+n2[i])

2
4
6
8
10


In [162]:
l1 = [1,2,3,4]
l2 = [3,4,6,7]
l1.append(l2)
l1

[1, 2, 3, 4, [3, 4, 6, 7]]

In [169]:
fun  = lambda x:x*x
fun(5)

25

# Task:

* create one excel.. 
Std_name Eng Maths
a        12   35
b        45   67
c        56   78
* read in Dataframe
Std_name Eng Maths  Total 
a        12   35     47
b        45   67     102
c        56   78     235

* export to report.csv